File: examples/azure-functions/csharp/Services/DatabricksClient.cs
using System.Net.Http.Json;
using System.Text.Json;
using ExcelDatabricksFunction.Models;

namespace ExcelDatabricksFunction.Services;

/// <summary>
/// Databricks SQL Statement Execution client interface.
/// </summary>
public interface IDatabricksClient
{
    Task<StatementResult> ExecuteStatementAsync(
        string statement,
        Dictionary<string, string>? parameters = null,
        string waitTimeout = "50s",
        string format = "JSON_ARRAY",
        CancellationToken cancellationToken = default);

    Task<StatementResult> GetStatementStatusAsync(
        string statementId,
        CancellationToken cancellationToken = default);

    Task<StatementResult> GetChunkAsync(
        string statementId,
        int chunkIndex,
        CancellationToken cancellationToken = default);

    Task<bool> CancelStatementAsync(
        string statementId,
        CancellationToken cancellationToken = default);
}

/// <summary>
/// Databricks SQL Statement Execution client.
///
/// Uses IHttpClientFactory for connection pooling to prevent SNAT port exhaustion.
///
/// Reference:
/// - https://docs.databricks.com/api/workspace/statementexecution
/// - https://learn.microsoft.com/en-us/azure/azure-functions/manage-connections
/// </summary>
public class DatabricksClient : IDatabricksClient
{
    private readonly HttpClient _httpClient;
    private readonly string _warehouseId;
    private readonly JsonSerializerOptions _jsonOptions;

    public DatabricksClient(IHttpClientFactory httpClientFactory)
    {
        _httpClient = httpClientFactory.CreateClient("Databricks");
        _warehouseId = Environment.GetEnvironmentVariable("DATABRICKS_WAREHOUSE_ID")
            ?? throw new InvalidOperationException("DATABRICKS_WAREHOUSE_ID not configured");

        _jsonOptions = new JsonSerializerOptions
        {
            PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower
        };
    }

    public async Task<StatementResult> ExecuteStatementAsync(
        string statement,
        Dictionary<string, string>? parameters = null,
        string waitTimeout = "50s",
        string format = "JSON_ARRAY",
        CancellationToken cancellationToken = default)
    {
        var payload = new Dictionary<string, object>
        {
            ["warehouse_id"] = _warehouseId,
            ["statement"] = statement,
            ["wait_timeout"] = waitTimeout,
            ["on_wait_timeout"] = "CONTINUE",
            ["format"] = format
        };

        if (parameters != null)
        {
            payload["parameters"] = parameters.Select(p => new { name = p.Key, value = p.Value }).ToList();
        }

        var response = await _httpClient.PostAsJsonAsync(
            "/api/2.0/sql/statements",
            payload,
            cancellationToken);

        response.EnsureSuccessStatusCode();

        var data = await response.Content.ReadFromJsonAsync<DatabricksResponse>(
            _jsonOptions, cancellationToken);

        return ParseResponse(data!);
    }

    public async Task<StatementResult> GetStatementStatusAsync(
        string statementId,
        CancellationToken cancellationToken = default)
    {
        var response = await _httpClient.GetAsync(
            $"/api/2.0/sql/statements/{statementId}",
            cancellationToken);

        response.EnsureSuccessStatusCode();

        var data = await response.Content.ReadFromJsonAsync<DatabricksResponse>(
            _jsonOptions, cancellationToken);

        return ParseResponse(data!);
    }

    public async Task<StatementResult> GetChunkAsync(
        string statementId,
        int chunkIndex,
        CancellationToken cancellationToken = default)
    {
        var response = await _httpClient.GetAsync(
            $"/api/2.0/sql/statements/{statementId}/result/chunks/{chunkIndex}",
            cancellationToken);

        response.EnsureSuccessStatusCode();

        var data = await response.Content.ReadFromJsonAsync<DatabricksChunkResponse>(
            _jsonOptions, cancellationToken);

        return new StatementResult
        {
            StatementId = statementId,
            State = StatementState.SUCCEEDED,
            ChunkIndex = data?.ChunkIndex ?? chunkIndex,
            Data = data?.DataArray
        };
    }

    public async Task<bool> CancelStatementAsync(
        string statementId,
        CancellationToken cancellationToken = default)
    {
        try
        {
            var response = await _httpClient.PostAsync(
                $"/api/2.0/sql/statements/{statementId}/cancel",
                null,
                cancellationToken);

            return response.IsSuccessStatusCode;
        }
        catch
        {
            return false;
        }
    }

    private static StatementResult ParseResponse(DatabricksResponse data)
    {
        var stateStr = data.Status?.State ?? "PENDING";
        var state = Enum.TryParse<StatementState>(stateStr, out var parsed)
            ? parsed
            : StatementState.PENDING;

        List<Column>? columns = null;
        if (data.Manifest?.Schema?.Columns != null)
        {
            columns = data.Manifest.Schema.Columns
                .Select(c => new Column(c.Name ?? "", c.TypeName ?? "string"))
                .ToList();
        }

        return new StatementResult
        {
            StatementId = data.StatementId ?? "",
            State = state,
            TotalChunkCount = data.Manifest?.TotalChunkCount,
            TotalRowCount = data.Manifest?.TotalRowCount,
            Columns = columns,
            Data = data.Result?.DataArray,
            ChunkIndex = data.Result?.ChunkIndex,
            ErrorMessage = state == StatementState.FAILED
                ? data.Status?.Error?.Message ?? "Unknown error"
                : null
        };
    }
}

File: examples/azure-functions/typescript/src/services/databricksClient.ts
/**
 * Databricks SQL Statement Execution Client
 *
 * Uses native fetch with connection reuse to avoid SNAT port exhaustion.
 *
 * Reference:
 * - https://docs.databricks.com/api/workspace/statementexecution
 * - https://learn.microsoft.com/en-us/azure/azure-functions/manage-connections
 */

// =============================================================================
// Types
// =============================================================================

export enum StatementState {
  PENDING = 'PENDING',
  RUNNING = 'RUNNING',
  SUCCEEDED = 'SUCCEEDED',
  FAILED = 'FAILED',
  CANCELED = 'CANCELED',
  CLOSED = 'CLOSED',
}

export interface Column {
  name: string;
  type: string;
}

export interface StatementResult {
  statementId: string;
  state: StatementState;
  totalChunkCount?: number;
  totalRowCount?: number;
  columns?: Column[];
  data?: unknown[][];
  chunkIndex?: number;
  errorMessage?: string;
}

interface DatabricksConfig {
  host: string;
  token: string;
  warehouseId: string;
  timeout?: number;
}

interface DatabricksResponse {
  statement_id: string;
  status: {
    state: string;
    error?: { message: string };
  };
  manifest?: {
    total_chunk_count: number;
    total_row_count: number;
    schema?: {
      columns: Array<{ name: string; type_name?: string }>;
    };
  };
  result?: {
    chunk_index: number;
    data_array?: unknown[][];
  };
}

// =============================================================================
// Client
// =============================================================================

export class DatabricksClient {
  private readonly host: string;
  private readonly token: string;
  private readonly warehouseId: string;
  private readonly timeout: number;

  constructor(config?: Partial<DatabricksConfig>) {
    this.host = (config?.host ?? process.env.DATABRICKS_HOST ?? '').replace(/\/$/, '');
    this.token = config?.token ?? process.env.DATABRICKS_TOKEN ?? '';
    this.warehouseId = config?.warehouseId ?? process.env.DATABRICKS_WAREHOUSE_ID ?? '';
    this.timeout = config?.timeout ?? 60000;

    if (!this.host || !this.token || !this.warehouseId) {
      throw new Error(
        'Missing Databricks config. Set DATABRICKS_HOST, DATABRICKS_TOKEN, ' +
        'and DATABRICKS_WAREHOUSE_ID environment variables.'
      );
    }
  }

  /**
   * Execute a SQL statement and return initial results.
   */
  async executeStatement(
    statement: string,
    parameters?: Record<string, string>,
    waitTimeout = '50s',
    format = 'JSON_ARRAY'
  ): Promise<StatementResult> {
    const payload: Record<string, unknown> = {
      warehouse_id: this.warehouseId,
      statement,
      wait_timeout: waitTimeout,
      on_wait_timeout: 'CONTINUE',
      format,
    };

    if (parameters) {
      payload.parameters = Object.entries(parameters).map(([name, value]) => ({
        name,
        value: String(value),
      }));
    }

    const response = await this.fetch('/api/2.0/sql/statements', {
      method: 'POST',
      body: JSON.stringify(payload),
    });

    return this.parseResponse(response);
  }

  /**
   * Get current status of a statement execution.
   */
  async getStatementStatus(statementId: string): Promise<StatementResult> {
    const response = await this.fetch(`/api/2.0/sql/statements/${statementId}`);
    return this.parseResponse(response);
  }

  /**
   * Get a specific result chunk by index.
   */
  async getChunk(statementId: string, chunkIndex: number): Promise<StatementResult> {
    const response = await this.fetch(
      `/api/2.0/sql/statements/${statementId}/result/chunks/${chunkIndex}`
    );

    return {
      statementId,
      state: StatementState.SUCCEEDED,
      chunkIndex: response.chunk_index ?? chunkIndex,
      data: response.data_array,
    };
  }

  /**
   * Cancel a running statement.
   */
  async cancelStatement(statementId: string): Promise<boolean> {
    try {
      await this.fetch(`/api/2.0/sql/statements/${statementId}/cancel`, {
        method: 'POST',
      });
      return true;
    } catch {
      return false;
    }
  }

  private async fetch(path: string, options?: RequestInit): Promise<DatabricksResponse> {
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeout);

    try {
      const response = await fetch(`${this.host}${path}`, {
        ...options,
        headers: {
          'Authorization': `Bearer ${this.token}`,
          'Content-Type': 'application/json',
          ...options?.headers,
        },
        signal: controller.signal,
      });

      if (!response.ok) {
        const text = await response.text();
        throw new Error(`Databricks API error ${response.status}: ${text}`);
      }

      return response.json() as Promise<DatabricksResponse>;
    } finally {
      clearTimeout(timeoutId);
    }
  }

  private parseResponse(data: DatabricksResponse): StatementResult {
    const state = data.status.state as StatementState;

    const result: StatementResult = {
      statementId: data.statement_id,
      state,
    };

    if (state === StatementState.FAILED) {
      result.errorMessage = data.status.error?.message ?? 'Unknown error';
    }

    if (data.manifest) {
      result.totalChunkCount = data.manifest.total_chunk_count;
      result.totalRowCount = data.manifest.total_row_count;

      if (data.manifest.schema?.columns) {
        result.columns = data.manifest.schema.columns.map((c) => ({
          name: c.name,
          type: c.type_name ?? 'string',
        }));
      }
    }

    if (data.result) {
      result.chunkIndex = data.result.chunk_index;
      result.data = data.result.data_array;
    }

    return result;
  }
}

// =============================================================================
// Singleton
// =============================================================================

let clientInstance: DatabricksClient | null = null;

/**
 * Get or create the singleton Databricks client.
 * Ensures connection reuse across all function invocations.
 */
export function getDatabricksClient(): DatabricksClient {
  if (!clientInstance) {
    clientInstance = new DatabricksClient();
  }
  return clientInstance;
}
